src/ioctls/common/kvm_vec.rs

```diff
@@ -0,0 +1,549 @@
+///         &self.entries
+///     }
+///
+///     fn entries_mut(&mut self) -> &mut __IncompleteArrayField<u32> {
+///         &mut self.entries
+///     }
+/// }
+/// ```
+#[allow(clippy::len_without_is_empty)]
+pub trait KvmArray {
+    type Entry: PartialEq + Copy;
+
+    /// Get the array length
+    ///
+    fn len(&self) -> usize;
+
+    /// Get the array length as mut
+    ///
+    fn set_len(&mut self, len: usize);
+
+    /// Get max array length
+    ///
+    fn max_len() -> usize;
+
+    /// Get the array entries
+    ///
+    fn entries(&self) -> &__IncompleteArrayField<Self::Entry>;
+
+    /// Get the array entries as mut
+    ///
+    fn entries_mut(&mut self) -> &mut __IncompleteArrayField<Self::Entry>;
+}
+
+/// An adapter that helps in treating a KvmArray similarly to an actual `Vec`.
+///
+pub struct KvmVec<T: Default + KvmArray> {
+    // this variable holds the KvmArray structure. We use a `Vec<T>` To make the allocation
+    // large enough while still being aligned for `T`. Only the first element of `Vec<T>` will
+    // actually be used as a `T`. The remaining memory in the `Vec<T>` is for `entries`, which
+    // must be contiguous. Since the entries are of type `KvmArray::Entry`
+    // we must be careful to convert the desired capacity of the `KvmVec`
+    // from `KvmArray::Entry` to `T` when reserving or releasing memory.
+    mem_allocator: Vec<T>,
+    // the number of elements of type `KvmArray::Entry` currently in the vec
+    len: usize,
+    // the capacity of the `KvmVec` measured in elements of type `KvmArray::Entry`
+    capacity: usize,
+}
+
+impl<T: Default + KvmArray> KvmVec<T> {
+    /// Get the capacity required by mem_allocator in order to hold
+    /// the provided number of  `KvmArray::Entry`
+    ///
+    fn kvm_vec_len_to_mem_allocator_len(kvm_vec_len: usize) -> usize {
+        let kvm_vec_size_in_bytes = size_of::<T>() + kvm_vec_len * size_of::<T::Entry>();
+        (kvm_vec_size_in_bytes + size_of::<T>() - 1) / size_of::<T>()
+    }
+
+    /// Get the number of elements of type `KvmArray::Entry` that fit
+    /// in a mem_allocator of provided len
+    ///
+    fn mem_allocator_len_to_kvm_vec_len(mem_allocator_len: usize) -> usize {
+        if mem_allocator_len == 0 {
+            return 0;
+        }
+
+        let array_size_in_bytes = (mem_allocator_len - 1) * size_of::<T>();
+        array_size_in_bytes / size_of::<T::Entry>()
+    }
+
+    /// Constructs a new KvmVec<T> that contains `num_elements` empty elements
+    /// of type `KvmArray::Entry`
+    ///
+    /// # Arguments
+    ///
+    /// * `num_elements` - The number of empty elements of type `KvmArray::Entry` in the initial `KvmVec`
+    ///
+    pub fn new(num_elements: usize) -> KvmVec<T> {
+        let required_mem_allocator_capacity =
+            KvmVec::<T>::kvm_vec_len_to_mem_allocator_len(num_elements);
+
+        let mut mem_allocator = Vec::with_capacity(required_mem_allocator_capacity);
```

Comment:
I think lines 138-141 can be replaced with `mem_allocator = vec![T::default(); required_mem_allocator_capacity]`, if it looks any better.